# 1º Semana

Os dados passados foram: 

[Base de Dados - Alura Voz](https://github.com/sthemonica/alura-voz/blob/main/Dados/Telco-Customer-Churn.json)

[Dicionário de Dados](https://github.com/sthemonica/alura-voz/blob/main/dicionario.md)


### Leitura do Arquivo JSON

In [1]:
import pandas as pd

In [2]:
dataset = pd.read_json(r'C:\Users\Jose Eduardo\Documents\GitHub\alura_challenge\Telco-Customer-Churn.json')
dataset.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

Verificando o arquivo 'Telco-Customer-Churn.json' entendemos que as colunas: customer, phone, internet, account tem mais dados que precisaremos analizar, iremos separar todos esses dados para entender qual o melhor caminho tomar.

### Separando customer, phone, internet, account

#### Dados Coluna 'customer'

In [3]:
dataset_customer = pd.json_normalize(dataset.customer)
dataset_customer.head(3)

gender  SeniorCitizen Partner Dependents  tenure
0  Female              0     Yes        Yes       9
1    Male              0      No         No       9
2    Male              0      No         No       4

#### Dados Coluna 'phone'

In [4]:
dataset_phone = pd.json_normalize(dataset.phone)
dataset_phone.head(3)

PhoneService MultipleLines
0          Yes            No
1          Yes           Yes
2          Yes            No

#### Dados Coluna 'internet'

In [5]:
dataset_internet = pd.json_normalize(dataset.internet)
dataset_internet.head(3)

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No         Yes   
1             DSL             No           No               No          No   
2     Fiber optic             No           No              Yes          No   

  StreamingTV StreamingMovies  
0         Yes              No  
1          No             Yes  
2          No              No

#### Dados Coluna 'account'

In [6]:
dataset_account = pd.json_normalize(dataset.account)
dataset_account.head(3)

Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0        One year              Yes      Mailed check             65.6   
1  Month-to-month               No      Mailed check             59.9   
2  Month-to-month              Yes  Electronic check             73.9   

  Charges.Total  
0         593.3  
1         542.4  
2        280.85

#### Unindo os DataFrames

Agora que temos todas as colunas separadas, entendemos que no nosso DATASET principal, as colunas: customer, phone, internet, account tem muito mais dados que precisaremos verificar, dessa forma o melhor a se fazer é unir as colunas que separamos acima. Para fazer isso iremos criar um novo DataFrame que terá as colunas: 'customerID', 'Churn' do DataFrame original e os 4 DataFrames que separamos acima. Para isso utilizaremos o método concat do Pandas.

In [7]:
dataset_concat = pd.concat([dataset[['customerID', 'Churn']],dataset_customer, dataset_phone, dataset_internet, dataset_account], axis=1)
dataset_concat.head(3)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  

[3 rows x 21 columns]

Como estou estudando achei interessante, além de concatenar os dados com o .concat(), utilizar também o .merge() para fixar melhor o conhecimento. Agora temos um UNICO DataFrame com todos os dados que iremos análisar.

In [8]:
dataset_merge_customer_phone = pd.merge(
    left = dataset_customer,
    right = dataset_phone,
    left_index = True,
    right_index = True
)
dataset_merge_customer_phone.head(3)

gender  SeniorCitizen Partner Dependents  tenure PhoneService MultipleLines
0  Female              0     Yes        Yes       9          Yes            No
1    Male              0      No         No       9          Yes           Yes
2    Male              0      No         No       4          Yes            No

In [9]:
dataset_merge_internet_account = pd.merge(
    left= dataset_internet,
    right= dataset_account,
    left_index=True,
    right_index=True
)
dataset_merge_internet_account.head(3)

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No         Yes   
1             DSL             No           No               No          No   
2     Fiber optic             No           No              Yes          No   

  StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes              No        One year              Yes   
1          No             Yes  Month-to-month               No   
2          No              No  Month-to-month              Yes   

      PaymentMethod  Charges.Monthly Charges.Total  
0      Mailed check             65.6         593.3  
1      Mailed check             59.9         542.4  
2  Electronic check             73.9        280.85

In [10]:
dataset_merge = pd.merge(left=dataset_merge_customer_phone, right=dataset_merge_internet_account, left_index=True, right_index=True)
dataset_merge.head(3)

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes        Yes       9          Yes   
1    Male              0      No         No       9          Yes   
2    Male              0      No         No       4          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0            No             DSL             No          Yes               No   
1           Yes             DSL             No           No               No   
2            No     Fiber optic             No           No              Yes   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes         Yes              No        One year              Yes   
1          No          No             Yes  Month-to-month               No   
2          No          No              No  Month-to-month              Yes   

      PaymentMethod  Charges.Monthly Charges.Total  
0      Mailed check             65.6         593.3  
1      Mailed check             59.9         542.4  
2  Electronic check             73.9        280.85

In [11]:
dataset_concat.head(3)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  

[3 rows x 21 columns]

### Entender quais informações o conjunto de dados possui

Nessa etapa do projeto já unificamos todos os DataFrames, agora precisamos usar o [Dicionário de Dados](https://github.com/sthemonica/alura-voz/blob/main/dicionario.md) para entender do que se trata cada uma das colunas.

<img src="didionário de dados.png">

Análisando o didionário de dados entendemos o que cada uma das colunas do nosso dataset se tratam, com isso podemos ir para o próximo passo que será verificar quais os TIPOS DE DADOS do nosso dataset.

### Analisar quais os tipos de dados

Para iniciarmos a análise dos tipos de dados usaremos a função .info() em nosso DataFrame.

In [12]:
dataset_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


Analisando as informações recebidas entendemos que nenhuma das colunas tem dados 'NULOS', na coluna SeniorCitizen está como inteiro, a coluna tenure está como inteiro, a coluna Charges.Monthly está como float. Analisando os dados da coluna Charges.Total verificamos que essa coluna se trata do TOTAL GASTO PELO CLIENTE, ou seja Charges.Monthly * tenure, logo encontramos uma falha, pois essa coluna também deveria ser uma coluna do tipo float. Agora precisamos identificar possiveis inconcistencias nos dados de cada coluna.

### Verificar quais são as inconsistências nos dados

Para verificar possiveis inconsistencias faremos um loop para verificar todas as 21 colunas de nosso DataFrame conforme abaixo:

In [13]:
for data in dataset_concat:
    print('='*50)
    print()
    print(dataset_concat[data].value_counts())
    print()


0002-ORFBO    1
6614-VBEGU    1
6637-KYRCV    1
6635-MYYYZ    1
6635-CPNUN    1
             ..
3374-TTZTK    1
3374-PZLXD    1
3374-LXDEV    1
3373-YZZYM    1
9995-HOTOH    1
Name: customerID, Length: 7267, dtype: int64


No     5174
Yes    1869
        224
Name: Churn, dtype: int64


Male      3675
Female    3592
Name: gender, dtype: int64


0    6085
1    1182
Name: SeniorCitizen, dtype: int64


No     3749
Yes    3518
Name: Partner, dtype: int64


No     5086
Yes    2181
Name: Dependents, dtype: int64


1     634
72    369
2     246
3     207
4     185
     ... 
38     60
39     59
44     54
36     50
0      11
Name: tenure, Length: 73, dtype: int64


Yes    6560
No      707
Name: PhoneService, dtype: int64


No                  3495
Yes                 3065
No phone service     707
Name: MultipleLines, dtype: int64


Fiber optic    3198
DSL            2488
No             1581
Name: InternetService, dtype: int64


No                     3608
Yes                    2078
No internet

Com isso identificamos que na coluna Churn existem 224 registros VAZIOS, diferente de todas as colunas do tipo bool (Yes ou No) a coluna SeniorCitizen está com 0 para Não e 1 para Sim, na coluna tenure existem 11 registros como 0, na coluna Charges.Total existem 11 registros VAZIOS. 

### Corrigir as inconsistências nos dados

Conforme erros identificados acima devemos corrigi-los. Corrigiremos primeiro a coluna Churn, como o intuito dessa analise é criar a persona para o Churn iremos excluir os 224 registros vazios do nosso dataset, para isso utilizaremos a função .drop()

In [14]:
vazios = dataset_concat[dataset_concat['Churn'] == ''].index
dataset_concat.drop(vazios, inplace=True)

In [15]:
dataset_concat['Churn'].info()

<class 'pandas.core.series.Series'>
Int64Index: 7043 entries, 0 to 7266
Series name: Churn
Non-Null Count  Dtype 
--------------  ----- 
7043 non-null   object
dtypes: object(1)
memory usage: 110.0+ KB


Com isso excluimos os 224 registros da nossa coluna Churn, agora precisamos resetar o Index do dataset, para isso utilizaremos a função .reset_index()

In [16]:
dataset_concat.reset_index(inplace=True)

In [17]:
dataset_concat.drop(["index"], axis=1, inplace=True)

Agora a inconsistência na coluna Churn foi resolvida. O próximo passo é alterar os valores da coluna SeniorCitizen, para isso utilizaremos a função .replace()

In [18]:
dataset_concat['SeniorCitizen'].replace(0, 'No', inplace=True)
dataset_concat['SeniorCitizen'].replace(1, 'Yes', inplace=True)

In [19]:
dataset_concat['SeniorCitizen'].value_counts()

No     5901
Yes    1142
Name: SeniorCitizen, dtype: int64

Dessa forma a inconsistência na coluna SeniorCitizen foi resolvida.

Na coluna Charges.Total existem 11 registros VAZIOS, isso não faz sentido, pois na coluna Charges.Monthly os 11 registros estão preenchidos, o que devemos fazer é repetir o valor da coluna Charges.Monthly na coluna Charges.Total, para isso criaremos uma variavel do tipo INDEX e utilizaremos a função .loc() para acessar o grupo de linhas que tem o valor VAZIO e após isso utilizaremos a função .loc() para acessar o grupo de linhas de Charges.Total e atribuiremos o valor.

In [20]:
idx = dataset_concat[dataset_concat['Charges.Total'] == ' '].index
dataset_concat.loc[idx, "Charges.Total"] = dataset_concat.loc[idx, "Charges.Monthly"]

Agora se usarmos a função .query() para fazer uma consulta na coluna Charges.Total, Charges.Monthly e tenure conseguiremos verificar que os dados são iguais.

In [21]:
dataset_concat.query('tenure == 0')[['Charges.Total', 'Charges.Monthly', 'tenure']]

Charges.Total  Charges.Monthly  tenure
945          56.05            56.05       0
1731          20.0            20.00       0
1906          61.9            61.90       0
2025          19.7            19.70       0
2176         20.25            20.25       0
2250         25.35            25.35       0
2855         73.35            73.35       0
3052         25.75            25.75       0
3118         52.55            52.55       0
4054         80.85            80.85       0
5433         19.85            19.85       0

Conforme DataFrame acima foi resolvido a inconsistência na coluna Charges.Total. Agora podemos resolver a ultima inconsistêmcia que é o TIPO da coluna Charges.Total, iremos transformala em float com a função .astype(float).

In [22]:
dataset_concat['Charges.Total'] = dataset_concat['Charges.Total'].astype(float)

In [23]:
dataset_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   Churn             7043 non-null   object 
 2   gender            7043 non-null   object 
 3   SeniorCitizen     7043 non-null   object 
 4   Partner           7043 non-null   object 
 5   Dependents        7043 non-null   object 
 6   tenure            7043 non-null   int64  
 7   PhoneService      7043 non-null   object 
 8   MultipleLines     7043 non-null   object 
 9   InternetService   7043 non-null   object 
 10  OnlineSecurity    7043 non-null   object 
 11  OnlineBackup      7043 non-null   object 
 12  DeviceProtection  7043 non-null   object 
 13  TechSupport       7043 non-null   object 
 14  StreamingTV       7043 non-null   object 
 15  StreamingMovies   7043 non-null   object 
 16  Contract          7043 non-null   object 


Conforme acima a coluna Charges.Total agora também é do tipo float que é o correto para essa coluna.

### Traduzir Colunas

Para melhor entendimento, foi nos solicitado que traduzissemos as colunas para PT-BR


In [24]:
dataset_concat.columns = ['Identificação do Cliente', 
                          'Churn', 
                          'Sexo', 
                          'Idoso',
                          'Casado', 
                          'Dependentes', 
                          'Meses de Contrato', 
                          'Serviço Telefônico', 
                          'Multiplas Linhas', 
                          'Serviço de Internet',
                          'Proteção Internet',
                          'Backup Online',
                          'Proteção Dispositivo',
                          'Suporte Técnico',
                          'TV a Cabo',
                          'Stremaming Filmes',
                          'Tipo do Contrato',
                          'Fatura Online',
                          'Forma de Pagamento',
                          'Total Gasto Mensal',
                          'Total Gasto Anual'
                         ]
dataset_concat.head(3)

Identificação do Cliente Churn    Sexo Idoso Casado Dependentes  \
0               0002-ORFBO    No  Female    No    Yes         Yes   
1               0003-MKNFE    No    Male    No     No          No   
2               0004-TLHLJ   Yes    Male    No     No          No   

   Meses de Contrato Serviço Telefônico Multiplas Linhas Serviço de Internet  \
0                  9                Yes               No                 DSL   
1                  9                Yes              Yes                 DSL   
2                  4                Yes               No         Fiber optic   

   ... Backup Online Proteção Dispositivo Suporte Técnico TV a Cabo  \
0  ...           Yes                   No             Yes       Yes   
1  ...            No                   No              No        No   
2  ...            No                  Yes              No        No   

  Stremaming Filmes Tipo do Contrato Fatura Online Forma de Pagamento  \
0                No         One year           Yes       Mailed check   
1               Yes   Month-to-month            No       Mailed check   
2                No   Month-to-month           Yes   Electronic check   

  Total Gasto Mensal  Total Gasto Anual  
0               65.6             593.30  
1               59.9             542.40  
2               73.9             280.85  

[3 rows x 21 columns]

O próximo passo é traduzir todo o conteúdo do DataFrame, usaremos o .replace() para isso.

In [25]:
dataset_concat.replace('No', 'Não', inplace=True)
dataset_concat.replace('Yes', 'Sim', inplace=True)
dataset_concat.replace('Fiber optic', 'Fibra ótica', inplace=True)
dataset_concat.replace('Male', 'Masculino', inplace=True)
dataset_concat.replace('Female', 'Feminino', inplace=True)
dataset_concat.replace('No internet service', 'Não tem serviço de internet', inplace=True)
dataset_concat.replace('Month-to-month', 'Mês a mês', inplace=True)
dataset_concat.replace('Two year', 'Dois anos', inplace=True)
dataset_concat.replace('One year', 'Um ano', inplace=True)
dataset_concat.replace('Electronic check', 'Cheque eletrônico', inplace=True) 
dataset_concat.replace('Mailed check', 'Cheque', inplace=True)
dataset_concat.replace('Bank transfer (automatic)', 'Transferência bancária (automática)', inplace=True)
dataset_concat.replace('Credit card (automatic)', 'Cartão de crédito (automático)', inplace=True) 
dataset_concat.head(3)

Identificação do Cliente Churn       Sexo Idoso Casado Dependentes  \
0               0002-ORFBO   Não   Feminino   Não    Sim         Sim   
1               0003-MKNFE   Não  Masculino   Não    Não         Não   
2               0004-TLHLJ   Sim  Masculino   Não    Não         Não   

   Meses de Contrato Serviço Telefônico Multiplas Linhas Serviço de Internet  \
0                  9                Sim              Não                 DSL   
1                  9                Sim              Sim                 DSL   
2                  4                Sim              Não         Fibra ótica   

   ... Backup Online Proteção Dispositivo Suporte Técnico TV a Cabo  \
0  ...           Sim                  Não             Sim       Sim   
1  ...           Não                  Não             Não       Não   
2  ...           Não                  Sim             Não       Não   

  Stremaming Filmes Tipo do Contrato Fatura Online Forma de Pagamento  \
0               Não           Um ano           Sim             Cheque   
1               Sim        Mês a mês           Não             Cheque   
2               Não        Mês a mês           Sim  Cheque eletrônico   

  Total Gasto Mensal  Total Gasto Anual  
0               65.6             593.30  
1               59.9             542.40  
2               73.9             280.85  

[3 rows x 21 columns]

Com isso todo o conteúdo do DataFrame foi traduzido para Português Brasil.

### Inserir coluna contas diárias

Foi solicitado que criassemos uma nova Coluna na posição 18 com os valores gastos por dia. Para isso devemos pegar a coluna 'Total Gasto Mensal' e dividir por 30 que é a média dias no mês.

In [26]:
dataset_concat['Total Gasto Diário'] = dataset_concat['Total Gasto Mensal'] / 30

In [27]:
dataset_concat['Total Gasto Diário']

0       2.186667
1       1.996667
2       2.463333
3       3.266667
4       2.796667
          ...   
7038    1.838333
7039    2.836667
7040    1.676667
7041    2.261667
7042    1.966667
Name: Total Gasto Diário, Length: 7043, dtype: float64

In [28]:
dataset_concat.head(3)

Identificação do Cliente Churn       Sexo Idoso Casado Dependentes  \
0               0002-ORFBO   Não   Feminino   Não    Sim         Sim   
1               0003-MKNFE   Não  Masculino   Não    Não         Não   
2               0004-TLHLJ   Sim  Masculino   Não    Não         Não   

   Meses de Contrato Serviço Telefônico Multiplas Linhas Serviço de Internet  \
0                  9                Sim              Não                 DSL   
1                  9                Sim              Sim                 DSL   
2                  4                Sim              Não         Fibra ótica   

   ... Proteção Dispositivo Suporte Técnico TV a Cabo Stremaming Filmes  \
0  ...                  Não             Sim       Sim               Não   
1  ...                  Não             Não       Não               Sim   
2  ...                  Sim             Não       Não               Não   

  Tipo do Contrato Fatura Online Forma de Pagamento Total Gasto Mensal  \
0           Um ano           Sim             Cheque               65.6   
1        Mês a mês           Não             Cheque               59.9   
2        Mês a mês           Sim  Cheque eletrônico               73.9   

  Total Gasto Anual  Total Gasto Diário  
0            593.30            2.186667  
1            542.40            1.996667  
2            280.85            2.463333  

[3 rows x 22 columns]

Com isso finalizamos tudo que foi solicitado em nossa primeira semana.

### Criando novo arquivo CSV

Após fazer todas os ajustes no DataFrame, criarei um novo arquivo CSV para usar nas próximas análises.

In [29]:
dataset_concat.to_csv(r'C:\Users\Jose Eduardo\Documents\GitHub\alura_challenge\alura_challenge_1_semana.csv', index=False)